### PR TITLE
Fix: use --web-launch-url and --web-hostname arguments in flutter drive

### LINKED ISSUE #118028: --web-launch-url option has no effect for flutter drive
According to `flutter drive --help`, the option `--web-launch-url` is

> The URL to provide to the browser. Defaults to an HTTP URL with the host name of "--web-hostname", the port of "--web-port", and the path set to "/".

However, in my setting it has no effect.

## Steps to Reproduce

1. Create a new Flutter app.
2. Replace `main.dart` with the attached content.
3. Replace `pubspec.yaml` with the attached content.
4. Create `test_driver/integration_test.dart` with the attached content.
5. Create `integration_test/root_test.dart` with the attached content.
6. Create `integration_test/green_test.dart` with the attached content.
7. Start chromedriver on port 4444.
8. Run:

```bash
flutter drive --driver=test_driver/integration_test.dart --target=integration_test/green_test.dart -d chrome --web-port=43000 --web-launch-url='http://localhost:43000/#/green'
```

**Expected results:** The browser starts at `http://localhost:43000/#/green`, the green screen is shown, the test finds a widget with `/green` text and the test succeeds.

**Actual results:** The browser starts at `http://localhost:43000/#/`, the root screen is shown, the test fails to find a widget with `/green` text.

<details>
<summary>main.dart</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      routes: {
        '/': (context) => Scaffold(
              body: Container(
                child: Center(child: Text('root')),
              ),
            ),
        '/green': (context) => Scaffold(
              body: Container(
                color: Colors.green,
                child: Center(child: Text('/green')),
              ),
            ),
      },
    );
  }
}
```

</details>

<details>
<summary>pubspec.yaml</summary>

```yaml
name: integration_test_test
description: A new Flutter project.
publish_to: 'none'
version: 1.0.0+1

environment:
  sdk: '>=2.18.6 <3.0.0'

dependencies:
  flutter:
    sdk: flutter

dev_dependencies:
  flutter_test:
    sdk: flutter
  integration_test:
    sdk: flutter
```

</details>

<details>
<summary>test_driver/integration_test.dart</summary>

```dart
import 'package:integration_test/integration_test_driver.dart';

Future<void> main() => integrationDriver();
```

</details>

<details>
<summary>integration_test/root_test.dart</summary>

```dart
import 'package:flutter_test/flutter_test.dart';
import 'package:integration_test/integration_test.dart';

import 'package:integration_test_test/main.dart' as app;

void main() {
  IntegrationTestWidgetsFlutterBinding.ensureInitialized();

  testWidgets('Finds text in the root screen', (tester) async {
    app.main();
    await tester.pumpAndSettle();

    await Future.delayed(const Duration(seconds: 5)); // To see the screen.

    expect(find.text('root'), findsOneWidget);
  });
}
```

</details>
<details>
<summary>integration_test/green_test.dart</summary>

```dart
import 'package:flutter_test/flutter_test.dart';
import 'package:integration_test/integration_test.dart';

import 'package:integration_test_test/main.dart' as app;

void main() {
  IntegrationTestWidgetsFlutterBinding.ensureInitialized();

  testWidgets('Finds text in the green screen', (tester) async {
    app.main();
    await tester.pumpAndSettle();

    await Future.delayed(const Duration(seconds: 5)); // To see the screen.

    expect(find.text('/green'), findsOneWidget);
  });
}
```

</details>

## Screens

Video of the test:

https://user-images.githubusercontent.com/44893228/210737690-64c1c2cf-da22-4a5f-869c-714a08223f07.mp4

`http://localhost:43000/#/green` when running normally and typed in:

image

## Context

<details>
<summary>Ubuntu: flutter doctor -v</summary>

```
flutter doctor -v
[✓] Flutter (Channel stable, 3.3.10, on Ubuntu 20.04.5 LTS 5.14.0-1054-oem, locale en_US.UTF-8)
    • Flutter version 3.3.10 on channel stable at /home/alexey/snap/flutter/common/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 135454af32 (3 weeks ago), 2022-12-15 07:36:55 -0800
    • Engine revision 3316dd8728
    • Dart version 2.18.6
    • DevTools version 2.15.0

[!] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /home/alexey/Android/Sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/linux#android-setup for more details.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0-4ubuntu1
    • cmake version 3.16.3
    • ninja version 1.10.0
    • pkg-config version 0.29.1

[✓] Android Studio (version 2021.2)
    • Android Studio at /home/alexey/bin/android-studio
    • Flutter plugin version 71.0.2
    • Dart plugin version 212.5744
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] VS Code (version 1.74.2)
    • VS Code at /usr/share/code
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 20.04.5 LTS 5.14.0-1054-oem
    • Chrome (web)    • chrome • web-javascript • Google Chrome 108.0.5359.124

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
```

</details>

<details>
<summary>Ubuntu: chromedriver --version</summary>

```
ChromeDriver 108.0.5359.71 (1e0e3868ee06e91ad636a874420e3ca3ae3756ac-refs/branch-heads/5359@{#1016})
```

</details>

<details>
<summary>Mac: flutter doctor -v</summary>

```
➜  ~ flutter doctor -v
[✓] Flutter (Channel master, 3.7.0-14.0.pre.40, on macOS 12.6 21G115 darwin-x64, locale ru-RU)
    • Flutter version 3.7.0-14.0.pre.40 on channel master at /Users/dmitrijrepin/sdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision a562fe2d8a (4 hours ago), 2023-01-04 23:53:41 -0500
    • Engine revision b9b0193ea8
    • Dart version 3.0.0 (build 3.0.0-85.0.dev)
    • DevTools version 2.20.0

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0-rc2)
    • Android SDK at /Users/dmitrijrepin/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0-rc2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14B47b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)

[✓] IntelliJ IDEA Community Edition (version 2022.3)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.74.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.56.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 12.6 21G115 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 108.0.5359.124

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by deldering**
Implement expected functionalities when supplying `--web-launch-url` and/or `--web-hostname` arguments to `flutter drive`.
- `--web-launch-url` now sets the starting url for the (headless) browser
  - Which for example means you can start at a certain part of the app at the start of your integration test
- `--web-hostname` now sets the hostname where the target of flutter drive will be hosted
  - Which allows you to set something other than localhost (allowing access via a reverse-proxy for example)

Fixes #118028

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by andrewkolos** [comment]
@eyebrowsoffire Would you be interested in reviewing this PR?

File: packages/flutter_tools/lib/src/drive/web_driver_service.dart [hunk 1]
```diff
@@ -116,11 +121,16 @@ class WebDriverService extends DriverService {
       throw ToolExit('Failed to start application');
     }
 
-    _webUri = _residentRunner.uri;
-
-    if (_webUri == null) {
+    if (_residentRunner.uri == null) {
```
Comment (eyebrowsoffire) [inline packages/flutter_tools/lib/src/drive/web_driver_service.dart:124]: Should this check move down to after processing the debugging options? We only care want to error if we couldn't produce a valid `_webUri`, right?

**Review by eyebrowsoffire — APPROVED** [review]
Looks mostly good! One question/nitpick

File: packages/flutter_tools/lib/src/drive/web_driver_service.dart (hunk 1 above)
Comment (deldering-momo) [reply packages/flutter_tools/lib/src/drive/web_driver_service.dart:124]: I would say that both the `_webUri` and `_residentRunner.uri` should be valid. Even if you supply a different `webLaunchUrl` the original connection should still be able to be set up properly. That is why I didn't use `tryParse` for applying the `webLaunchUrl` later on if it is supplied.

Comment (deldering-momo) [reply packages/flutter_tools/lib/src/drive/web_driver_service.dart:124]: Another option would be to use `tryParse` if `webLaunchUrl` is available and throw a `ToolExit` if it returns `null`

Comment (eyebrowsoffire) [reply packages/flutter_tools/lib/src/drive/web_driver_service.dart:124]: I don't have a strong opinion on it. It's just a nitpick, and it's fine either way, I just think it makes more sense to validate the URI we're actually going to use. But I'm fine with landing it with any of those options.

File: dev/integration_tests/flutter_gallery/.gitignore [hunk 2]
```diff
@@ -1,3 +1,2 @@
-lib/generated_plugin_registrant.dart
```
Comment (christopherfujino) [inline dev/integration_tests/flutter_gallery/.gitignore:1]: why was this removed?

**Review by christopherfujino — APPROVED** [review]
LGTM

**Pull request merged**